lib/puma/thread_pool.rb

```diff
@@ -168,9 +168,17 @@ def reap
         dead_workers.each do |worker|
           worker.kill
           @spawned -= 1
+
+          # if there's work waiting for a usable thread and we now have available slots in our
+          # pool, spawn a thread to process that outstanding work
+          if @waiting < @todo.size and @spawned < @max
+            spawn_thread
+          end
         end
 
         @workers -= dead_workers
+
+        @not_empty.signal
```

Comment:
If all workers are dead, and there are no `@todo`s, then we _will_ be empty.  Is lying and signaling non-empty going to be a problem?

Reply:
Good question.  Signaling not empty when there is nothing waiting and no work to do is not a problem and it is consistent with the existing patterns of broadcast for that condition variable.  So, no problems here :)

Reply:
Ideal :)  Well then: looks great to me!